### PR TITLE
First stab at integrating Azure into the providers code

### PR DESCRIPTION
**Pull request by jteehan**
Better get this checked in.  Already lost it once.

Going to need help getting the right fields on the forms, but I think these are the fields azure will be using, at least in 5.6  I don't really care about 5.5

{{pytest: cfme/tests/cloud -k "test_provider_crud or test_instance_power_control" --use-provider azure }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 99992f9c0f164ad1a768b9896a927b674c338965:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
You have to have the latest credential yaml to see all the required values.
54 min ago	jteehan	Changed provision mode to incremental. Added subscript...

**Comment by dajoRH** [comment]
Lint report for commit 617e8f14a37c1273714358b6ed30b82422803aff:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9a2c6fd12542b46a7c665d55dc31979f7ad96bb7:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8d723e563b83a20e74dcd65992d3a807ae37da4b:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit c298be8494d332f6bc76efe4473b2dca79ba3c97:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@jteehan You might also take a peek whether https://github.com/RedHatQE/cfme_tests/blob/master/utils/virtual_machines.py#L41 does not need additional touches for azure.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by jteehan** [comment]
I don't know where to enter the credentials fields.  Basic form info is being entered okay.

**Comment by dajoRH** [comment]
Lint report for commit a6e02a2f3b1935633d10f791f260c080940f1834:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 07558dcce04f7e81d2a1e37de6a27bb029fe07ed:

`utils/providers.py`:
- :bangbang: Line 764:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 764:7: **E901** *SyntaxError: invalid syntax*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0ed8582ac09e1168d4385087cbba8033b1e574ea:

`cfme/cloud/provider.py`:
- :heavy_exclamation_mark: Line 176:1: **W293** *blank line contains whitespace*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by jteehan** [comment]
So, to see it working, you'll need to setup a local pytest as follows.  Little rough, but that's the easy part.

def test_create_provider_azure():
    azure_provider = cfme_data['management_systems']['azure']

    # Going to need this for tenant_id and subscription_id even though creds get's the other parts
    azure_creds = credentials[azure_provider['credentials']]

    creds = {}
    creds['Default'] = get_credentials_from_config('cloudqe_azure')

    azurep_args = {
        "name": 'TestNameABC',
        "credentials": creds,
        "zone": None,
        "key": None,
        "region": azure_provider['region'],
        "tenant_id": azure_creds['tenant_id'],
        "subscription_id": azure_creds['subscription_id']
    }

    az = AzureProvider(**azurep_args)
    az.create(cancel=False, validate_credentials=True)
    az.delete(cancel=False)
    az.wait_for_delete()

**Comment by dajoRH** [comment]
Lint report for commit 4f4727f10092bc826b99e4863bfda5350efec1ab:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
So I added the azure stuff to a few test files, but nothing runs.  Both say they don't work.  (mycfme)[jteehan@localhost cfme_tests]$ ../bin/py.test cfme/tests/cloud/test_instance_power_control.py 
collected 12 items 
collected 0 items (uncollected 12 items)

**Comment by dajoRH** [comment]
Lint report for commit 982897df7fe0a0d2759a377f6f4e30fe89df0d47:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@jteehan `--use-provider azure` ?

**Comment by mfalesni** [comment]
@jteehan btw you should not be running py.test directly as a binary, but rather do `../bin/activate` and then just do `py.test`.

File: cfme/tests/cloud/test_instance_power_control.py [hunk 1]
```diff
@@ -57,8 +58,12 @@ def check_power_options(soft_assert, instance, power_state):
     """ Checks if power options match given power state ('on', 'off')
     """
     must_be_available = {
+        AzureInstance: {
+            'on': [AzureInstance.STOP, AzureInstance.SOFT_REBOOT, AzureInstance.TERMINATE],
+            'off': [AzureInstance.START, AzureInstance.TERMINATE]
+        },
         EC2Instance: {
-            'on': [EC2Instance.STOP, EC2Instance.SOFT_REBOOT, EC2Instance.TERMINATE],
+            'on': [EC2Instance.STOP, EC2Instance.SUSPEND, EC2Instance.TERMINATE],
```
Comment (jkrocil) [inline cfme/tests/cloud/test_instance_power_control.py]: Is this a fix for EC2 or a mistake?

Comment (jteehan) [inline cfme/tests/cloud/test_instance_power_control.py]: No, Azure has two Stop modes.  We decided to change Stop to Stopped:Deallocate and Suspend to Stopped.  The latter doesn't stop the charges and was a bug fix.
https://bugzilla.redhat.com/show_bug.cgi?id=1316280
I don't think there is a reboot, but I can check.

Comment (jkrocil) [inline cfme/tests/cloud/test_instance_power_control.py]: All right, that sounds valid but you are editing expected EC2Instance states here, not Azure.

**Comment by dajoRH** [comment]
Lint report for commit 433789500f5b95dccd46b6ad9b6534bbcff55a65:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit dc0ef176a25df0b9debddc2797b1d771cc4a7164:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
As the senior lead on the team, please push it or explain to me why you don't understand and I'll explain it to  you, and then you can push it.

**Comment by mfalesni** [comment]
So, it seems the error was fixed somehow now. What is left to do for you is to rebase your branch over the latest master and push it. Then we'll see the proper results (I hope)

**Comment by mfalesni** [comment]
I looked in the PR tester code and it should be merging it on top of master so it might owrk properly now.

**Comment by dajohnso** [comment]
@jteehan rebase your branch over the latest master and we will try to re-run.  Fighting some sprout issues is not helping this effort so apologies and thank you for your continued (lack of) patience.  In terms of a lead on the team, not sure how that got in your head but please remove it.

**Comment by mfalesni** [comment]
@jteehan The code itself looks legit but it needs a rebase over the latest master and pushing so it can run with the fixes that should make the tests run properly again. Then I or anyone else in the team will be able to interpret the PR tester results.

I can't push on your branch.

**Comment by dajoRH** [comment]
Lint report for commit 8b3cdbaf548aa906fcefb73d6f5332c8456d1de1:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 3be83f2edfa471dcb8157fd7818894d9860acbb1:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit bfec00ec486e92c4397c29970f6245d86830b5de:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 69721dec6f981a92d7c97de02723e7ab29dd2ee5:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jteehan** [comment]
I found and fixed the 5.5 tenant_id issue.  As near as I can tell it added Azure provider to the appliance to both 5.5 and 5.6 but couldn't verify it even through it's clearly in the screen shot.  That should be easy enough to track down.

provider.validate_stats(ui=True)

How long does the appliance hang around for?  How can I go look manually.

**Pull request merged**